File: other-platforms/dp/count-subsets-with-sum-k.cpp
// https://www.naukri.com/code360/problems/number-of-subsets_3952532
// space optimised approach

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int findWays(vector<int> &arr, int k) {
        int n = arr.size();
        vector<int> prev(k + 1, 0);

        // write base cases
        prev[0] = 1;
        if(arr[0] <= k) {
            prev[arr[0]] = 1;
        }

        for(int idx = 1; idx < n; idx++) {
            vector<int> curr(k + 1, 0);
            curr[0] = 1;

            for(int target = 1; target <= k; target++) {
                int notTaken = prev[target];

                int taken = 0;
                if(arr[idx] <= target) {
                    taken = prev[target - arr[idx]];
                }

                curr[target] = taken + notTaken;
            }
            prev = curr;
        }
        return prev[k];
    }
};

int main() {
    Solution sol;
    
    vector<int> arr = {1, 2, 2, 3};
    int k = 3;

    cout << sol.findWays(arr, k) << endl;  // Output: 0 (empty not allowed)
    
    return 0;
}

File: other-platforms/dp/CountSubsetswithSumK.java
public class CountSubsetswithSumK {
    static int MOD = (int)(1e9 + 7);

    static int findWays(int[] num, int k) {
        int n = num.length;
        int[][] dp = new int[n][k + 1];

        // Base case
        if (num[0] == 0) {
            dp[0][0] = 2; // we can pick or not pick zero ? 2 ways
        } else {
            dp[0][0] = 1; // we can always make sum=0 by not picking
            if (num[0] <= k) {
                dp[0][num[0]] = 1;
            }
        }

        // Bottom-up DP
        for (int ind = 1; ind < n; ind++) {
            for (int target = 0; target <= k; target++) {
                int notTaken = dp[ind - 1][target];
                int taken = 0;
                if (num[ind] <= target) {
                    taken = dp[ind - 1][target - num[ind]];
                }
                dp[ind][target] = (notTaken + taken) % MOD;
            }
        }

        return dp[n - 1][k];
    }

    public static void main(String[] args) {
        int[] arr = {1, 1, 4, 5};
        int k = 5;
        
        System.out.println(findWays(arr, k));
    }
}
